crates/runtime/src/dataaccelerator/partitioned_duckdb/tables_mode/insert.rs

```diff
@@ -0,0 +1,196 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::{collections::HashMap, sync::Arc};
+
+use arrow::array::RecordBatch;
+use arrow_schema::SchemaRef;
+use async_stream::stream;
+use async_trait::async_trait;
+use datafusion::{
+    common::DFSchema,
+    error::DataFusionError,
+    logical_expr::dml::InsertOp,
+    physical_expr::create_physical_expr,
+    physical_plan::{
+        Distribution, ExecutionPlan, SendableRecordBatchStream, stream::RecordBatchStreamAdapter,
+    },
+    prelude::Expr,
+};
+use datafusion_datasource::sink::DataSinkExec;
+use datafusion_expr::execution_props::ExecutionProps;
+use datafusion_table_providers::{
+    duckdb::TableDefinition, sql::db_connection_pool::duckdbpool::DuckDbConnectionPool,
+    util::on_conflict::OnConflict,
+};
+use futures::StreamExt;
+use runtime_table_partition::{
+    expression::PartitionedBy,
+    insert::{InsertStrategy, PartitionContext, partition_batch},
+};
+
+use crate::{
+    dataaccelerator::partitioned_duckdb::tables_mode::sink::DuckDBPartitionedDataSink,
+    datafusion::extension::pass_thru::PassThruExec,
+};
+
+/// Strategy for handling `DuckDB` table-based partition insertions.
+#[derive(Debug)]
+pub struct DuckDBPartitionedInsertStrategy {
+    pool: Arc<DuckDbConnectionPool>,
+    table_definition: Arc<TableDefinition>,
+    on_conflict: Option<OnConflict>,
+}
+
+impl DuckDBPartitionedInsertStrategy {
+    #[must_use]
+    pub fn new(
+        pool: Arc<DuckDbConnectionPool>,
+        table_definition: Arc<TableDefinition>,
+        on_conflict: Option<OnConflict>,
+    ) -> Self {
+        Self {
+            pool,
+            table_definition,
+            on_conflict,
+        }
+    }
+
+    fn create_infer_partitions_exec(
+        input: Arc<dyn ExecutionPlan>,
+        ctx: &PartitionContext,
+    ) -> Arc<dyn ExecutionPlan> {
+        let creator = Arc::clone(&ctx.creator);
+        let partitions_lock = Arc::clone(&ctx.partitions);
+
+        let exec = move |input_exec: &Arc<dyn ExecutionPlan>, partition, ctx| {
+            let schema = input_exec.schema();
+            let input_stream = input_exec.execute(partition, ctx)?;
+            let creator = Arc::clone(&creator);
+            let partitions_lock = Arc::clone(&partitions_lock);
+
+            let s = stream! {
+                let mut input = input_stream;
+                let mut ok = true;
+                while let Some(item) = input.next().await {
+                    match item {
+                        Ok(b) => yield Ok(b),
+                        Err(e) => { ok = false; yield Err(e); }
+                    }
+                }
+                if ok {
+                    // Re-infer partitions and update context after successful completion
+                    match creator.infer_existing_partitions().await {
```

Comment:
What purpose does this serve? I would expect we know the partitions exactly after the insertion is complete and we don't need to infer it?